src/app/common/FabricTableImpl.h

```diff
@@ -0,0 +1,143 @@
+namespace chip {
+namespace app {
+namespace common {
+
+/**
+ * @brief Implementation of a storage in nonvolatile storage of a templatized table that stores by fabric id.
+ *
+ * DefaultSceneTableImpl is an implementation that allows to store scenes using PersistentStorageDelegate.
+ * It handles the storage of scenes by their ID, GroupID and EnpointID over multiple fabrics.
+ * It is meant to be used exclusively when the scene cluster is enable for at least one endpoint
+ * on the device.
+ */
+template <class StorageId, class StorageData, size_t kIteratorsMax>
+class FabricTableImpl
+{
+    using TableEntry    = data::TableEntry<StorageId, StorageData>;
+    using EntryIterator = CommonIterator<TableEntry>;
+    using EntryIndex    = data::EntryIndex;
+
+public:
+    virtual ~FabricTableImpl() { Finish(); };
+
+    CHIP_ERROR Init(PersistentStorageDelegate * storage);
+    void Finish();
+
+    // Scene count
+    CHIP_ERROR GetEndpointEntryCount(uint8_t & entry_count);
+    CHIP_ERROR GetFabricEntryCount(FabricIndex fabric_index, uint8_t & entry_count);
+
+    // Data
+    CHIP_ERROR GetRemainingCapacity(FabricIndex fabric_index, uint8_t & capacity);
+    CHIP_ERROR SetTableEntry(FabricIndex fabric_index, const TableEntry & entry);
+    CHIP_ERROR GetTableEntry(FabricIndex fabric_index, StorageId entry_id, TableEntry & entry);
+    CHIP_ERROR RemoveTableEntry(FabricIndex fabric_index, StorageId entry_id);
+    CHIP_ERROR RemoveTableEntryAtPosition(EndpointId endpoint, FabricIndex fabric_index, EntryIndex entry_idx);
+
+    // Fabrics
+    CHIP_ERROR RemoveFabric(FabricIndex fabric_index);
+    CHIP_ERROR RemoveEndpoint();
+
+    // Iterators
+    EntryIterator * IterateTableEntries(FabricIndex fabric_index);
+
+    void SetEndpoint(EndpointId endpoint);
+    void SetTableSize(uint16_t endpointEntryTableSize);
+    bool IsInitialized() { return (mStorage != nullptr); }
+
+protected:
+    // gcc bug prevents us from using a static variable; see:
+    // https://stackoverflow.com/questions/50638053/constexpr-static-data-member-without-initializer
+    static constexpr EndpointId kEntryEndpointClusterID();
+
+    // This constructor is meant for test purposes, it allows to change the defined max for entries per fabric and global, which
+    // allows to simulate OTA where this value was changed
+    FabricTableImpl(uint16_t maxEntriesPerFabric, uint16_t maxEntriesPerEndpoint) :
+        mMaxPerFabric(maxEntriesPerFabric), mMaxPerEndpoint(maxEntriesPerEndpoint)
+    {}
+
+    // Global entry count
+    CHIP_ERROR SetEndpointEntryCount(const uint8_t & entry_count);
+
+    // wrapper function around emberAfGetClustersFromEndpoint to allow override when testing
+    virtual uint8_t GetClustersFromEndpoint(ClusterId * clusterList, uint8_t listLen);
+
+    // wrapper function around emberAfGetClusterCountForEndpoint to allow override when testing
+    virtual uint8_t GetClusterCountFromEndpoint();
+
+    class EntryIteratorImpl : public EntryIterator
```

Comment:
As pointed out in my comment in SceneTableImp, it seems that the EntryIterator was made redundant when I implemented the entry_map. Since this is going to be a common object now, I am would it be better to remove the EntryIterator altogether or to at least have it in an external file so it can be used through composition instead of inheritance?

Reply:
I will defer to you, but `IterateTableEntries`, which instantiates the `EntryIterator`, has access to the class internals, including `FabricEntryData` accessed from the iterator, which is a hidden impl class...

Though there are certainly ways to tease that apart, the iterator does seem closely tied to the impl, so it may be a bit awkward

Is there a reason you're worried about having an iterator impl?  As long as it's not used, it should be optimized away during linking, right?

Reply:
That's something that I cannot confirm at the moment, I would double check.

Since we are declaring an ObjectPool of:
```
ObjectPool<EntryIteratorImpl, kIteratorsMax> mEntryIterators;
```
And we have the Finnish method making a call on it:
```
template <class StorageId, class StorageData, size_t kIteratorsMax>
void FabricTableImpl<StorageId, StorageData, kIteratorsMax>::Finish()
{
    mEntryIterators.ReleaseAll();
}
```
I am not sure what the compiler would deduce from this. I would at least make sure before keeping it it. I do agree that my initial idea of having it as a component wouldn't work though.

Reply:
Hm you're right, sorry somehow I missed the `mEntryIterators` field which of course will force it to be compiled & cannot be optimized away

I'll go ahead & remove it if you think it's redundant; I'll leave the class impl just in case someone does want to use it (that's the part that will be optimized away by linker), but I've removed `mEntryIterators` & references to it.

Reply:
There were test that were specifically verifying that the iterator worked, so these can be deleted. 

Else what we could do is create a TestSceneTableImpl that has a mEntryIterators member and insert it in unit tests.

Reply:
Sounds good, deleted & all tests passing now